### ISSUE TITLE
HTTP_PROXY proxy settings

### ISSUE THREAD

**Issue by aleksasiriski** [k3s-io/helm-controller#252]
Is it possible to configure proxy settings? I am using k3s and possibly rke2 in future, but I don't want to set HTTP_PROXY but instead use CONTAINERD_HTTP_PROXY and that works for pulling images, but this then doesn't propagate to helm install jobs and I can't seem to find how to set this up.

**Comment by brandond** [comment]
Kubernetes pods intentionally do not inherit environment variables from the kubelet or container runtime. If you want these env vars to be set, you must do so manually in the pod spec, or by using a tool such as https://github.com/redhat-cop/podpreset-webhook

**Comment by brandond** [comment]
Any HTTP_PROXY related env vars set in the helm controller's environment will be passed into the helm job pods. For k3s and rke2, this means you must set that env var if you want the helm controller's job pods to use the proxy.

Why don't you want to set HTTP_PROXY?

**Comment by aleksasiriski** [comment]
> Why don't you want to set HTTP_PROXY?

It's not very well documented what it affects, since I only need it for pulling images and helm install pods I managed to workaround this by using CiliumEgressGatewayPolicy with label keys for helm install pods.

Could you send me a link on what is being affected by setting HTTP_PROXY instead of CONTAINERD_PROXY if that docs exist?

**Comment by brandond** [comment]
> what is being affected by setting HTTP_PROXY instead of CONTAINERD_PROXY

Everything, just as you'd expect. Kubelet, apiserver, scheduler, controller-manager, etcd, all the things that run in the main K3s process. Pods are not of course, because they have their own environments.

**Comment by aleksasiriski** [comment]
> > what is being affected by setting HTTP_PROXY instead of CONTAINERD_PROXY
> 
> Everything, just as you'd expect. Kubelet, apiserver, scheduler, controller-manager, etcd, all the things that run in the main K3s process. Pods are not of course, because they have their own environments.

Yep, that's what I thought. I don't want my kubeapi and everything all to be at the mercy of proxy's uptime.

So, in the end, I would need to deploy my own Helm controller in order to give it proper HTTP_PROXY env vars, there's no way around it?

**Comment by brandond** [comment]
The apiserver, controller-manager, and such don't actually go out to the internet for anything, and the cluster CIDRs and cluster domain are all automatically added to the NO_PROXY list (as your proxy is unlikely to have access to things running inside the cluster), so I doubt you'd actually run into any problems simply setting the HTTP_PROXY env var. You might want to add your node LAN CIDR and internal DNS zone to the NO_PROXY list as well, just to be sure.

Have you actually tried it or are you just expecting there would be problems?

**Comment by aleksasiriski** [comment]
> The apiserver, controller-manager, and such don't actually go out to the internet for anything, and the cluster CIDRs and cluster domain are all automatically added to the NO_PROXY list (as your proxy is unlikely to have access to things running inside the cluster), so I doubt you'd actually run into any problems simply setting the HTTP_PROXY env var. You might want to add your node LAN CIDR and internal DNS zone to the NO_PROXY list as well, just to be sure.

Oh that's good then.

> Have you actually tried it or are you just expecting there would be problems?

I had problems that my proxy is deployed inside the cluster and using NodePort service with `localhost:<port>` and that doesn't work within the helm install container, but is required for containerd for example.

**Comment by brandond** [comment]
I'm confused what the value proposition would be there - why make your nodes go through something running in the cluster, to get out to the internet?

**Comment by aleksasiriski** [comment]
> I'm confused what the value proposition would be there - why make your nodes go through something running in the cluster, to get out to the internet?

Because I have to apply some filtering, to limit what can be download by commands like curl / helm install for example. I agree that it's an unconventional situation, but the idea was to avoid having yet another set of servers just for the proxy if possible.

**Comment by aleksasiriski** [comment]
> I'm confused what the value proposition would be there - why make your nodes go through something running in the cluster, to get out to the internet?

I came across these comments of mine and now understand that I wans't as elaborate as I could be. My situation is this:

I use cilium inside the cluster to filter based on IP, DNS and even HTTP(s) all the traffic happening inside the cluster. Instead of doing a separate filtering outside the cluster (when somebody uses ssh to go into a node and runs commands for example) I wanted to bring all the traffic that was outside the cluster *inside* the cluster. I managed to do this by running [a small boring proxy](https://github.com/kube-hetzner/boringproxy) and setting `/etc/environment` to use that (localhost since it's exposed as a NodePort service). This same proxy is set for k3s to pull images, in order to apply my filtering and should be set to Helm install pods.

I managed to avoid setting the proxy for Helm install pods, since they are running as pods already and I can apply my Cilium NetworkPolicies directly to those pods. So this issue is resolved for me, although I still believe it would be beneficial for people to have the option to set the proxy to helm controller separately from other components as it is possible with containerd. Perhaps even every component should be possible to set separately.
